src/Functions/DateTimeTransforms.h

```diff
@@ -57,25 +60,51 @@ struct ZeroTransform
     static UInt16 execute(UInt16, const DateLUTImpl &) { return 0; }
 };
 
+template <DateTimeOverflowMode date_time_overflow_mode = default_date_time_overflow_mode>
 struct ToDateImpl
 {
     static constexpr auto name = "toDate";
 
     static UInt16 execute(const DecimalUtils::DecimalComponents<DateTime64> & t, const DateLUTImpl & time_zone)
     {
-        return static_cast<UInt16>(time_zone.toDayNum(t.whole));
+        return execute(t.whole, time_zone);
     }
+
     static UInt16 execute(Int64 t, const DateLUTImpl & time_zone)
     {
-        return UInt16(time_zone.toDayNum(t));
+        if constexpr (date_time_overflow_mode == DateTimeOverflowMode::SATURATE)
```

Comment:
I am sure complexity would be a lot lower but performance would be the same if `date_time_overflow_mode` was not a template parameter but a normal function argument to all four overloads of `execute()`. (see my other comment about the CPU branch predictor).

Reply:
I trier doing this in #40217, but as far as I understood, it was reverted partly because of bad performance. Now it doesn't result in performance reduction in 'default' mode -- when the setting == `ignore`. This way we can be sure that performance IS the same.

Regarding code complexity: I don't think that passing an additional argument to functions is more readable than what I did now. Also taking into account we already have a `scale` argument.

Reply:
Okay, I see. We can keep the templates then.